lib/core/include/authentication_plugin_framework.hpp

```diff
@@ -0,0 +1,155 @@
+#include "irods_stacktrace.hpp"
+#include "miscServerFunct.hpp"
+#include "msParam.h"
+#include "rcConnect.h"
+#include "rodsDef.h"
+
+#ifdef RODS_SERVER
+#include "rsAuthCheck.hpp"
+#include "rsAuthRequest.hpp"
+#endif // RODS_SERVER
+
+#include <sstream>
+#include <string>
+#include <iostream>
+#include <termios.h>
+#include <unistd.h>
+
+#include <boost/any.hpp>
+
+#include <nlohmann/json.hpp>
+#include <openssl/md5.h>
+#include <fmt/format.h>
+
+using json = nlohmann::json;
+
+namespace irods::experimental
+{
+    namespace auth
+    {
+        static const std::string flow_complete{"authentication_flow_complete"};
+
+        static const std::string next_operation{"next_operation"};
+    }
+
+    class authentication_base : public irods::plugin_base
+    {
+    public:
+        #define OPERATION(C, F) std::function<json(C*, const json&)>([&](C* c, const json& j) -> json {return F(c, j);})
+
+        authentication_base()
+            : plugin_base{"authentication_framework_plugin", "empty_context_string"}
+        {
+            add_operation(AUTH_CLIENT_START, OPERATION(rcComm_t, auth_client_start));
+        } // ctor
+
+        virtual json auth_client_start(rcComm_t* comm, const json& req) = 0;
+
+        template<typename RxComm>
+        void add_operation(const std::string& n, std::function<json(RxComm*, const json&)> f)
+        {
+            if(operations_.find(n) != operations_.end()) {
+                THROW(SYS_INTERNAL_ERR, fmt::format("operation already exists [{}]", n));
+            }
+
+            if(n.empty()) {
+                THROW(SYS_INVALID_INPUT_PARAM, fmt::format("operation name is empty [{}]", n));
+            }
+
+            operations_[n] = f;
+        } // add_operation
+
+        template<typename RxComm>
+        json call(RxComm* comm, const std::string& n, const json& req)
+        {
+            auto itr = operations_.find(n);
+            if(itr == operations_.end()) {
+                THROW(SYS_INVALID_INPUT_PARAM,
+                      fmt::format("call operation :: missing operation[{}]", n));
+            }
+
+            using fcn_t = std::function<json(RxComm*, const json&)>;
+            auto op = boost::any_cast<fcn_t&>(operations_[n]);
+
+            return op(comm, req);
+        } // call
+    }; // class authentication_base
+
+    auto resolve_authentication_plugin(const std::string& scheme, const std::string& type)
+    {
+        using plugin_type = irods::experimental::authentication_base;
+
+        std::string lower = scheme;
+        std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);
+
+        const auto name = fmt::format("irods_auth_plugin-{}_{}", lower, type);
+
+        plugin_type* plugin{};
+        auto err = irods::load_plugin<plugin_type>(
+                        plugin,
+                        name,
+                        irods::PLUGIN_TYPE_AUTHENTICATION,
+                        "irods::experimental::authentication_base",
```

Comment:
Looking at this again... I don't know that this string actually used for this plugin interface. The `load_plugin` call is attempting to load a factory method from the plugin's shared object. In this case, the factory method is provided in the auth plugins (see irods_auth_plugin-native.cpp). So, this string could actually be anything because the two `const std::string&` parameters passed to the factory method in the plugin are unused (again, see the native plugin implementation).

After a quick look around, it looks like most of the plugins are not using these parameters, but it is part of the interface for constructing the `irods::plugin_base`. We are not using that in this instance, so I think this string can be whatever.

Reply:
What do you think about using the string "authentication_plugin_base" here? We could also just pass an empty string since it's not being used for these plugins. None of the other plugins are namespacing, but maybe there was a purpose for this in the original implementation?

Reply:
if nothing is using it, maybe we can re-purpose it for debugging or something else.

i don't think we should pass an empty string. the string you suggested works. does it appear in the log file?

alternative ideas:
- `"authentication_plugin_framework"`: indicator for what the plugin is related to
- build a string using info about the target plugin

Reply:
Hmm... I dug... again... and found that this is ultimately being piped into the `instance_name_` member of the `plugin_base` in instances where this is being used (e.g. the database plugin). So maybe we could just use the `name` of the plugin for this parameter as well, which would be bullet 2 from your suggestions. Thoughts?

Reply:
sounds fine to me.